posts/2021/09/re-engineering-cicd-pipelines-for-scipy.md

```diff
@@ -0,0 +1,209 @@
+
+- macOS (X86): Supported on a GitHub Action workflow for pull requests.
+
+- macOS (ARM64): Requires more effort with no options available on GitHub Actions.
+
+We also implemented a sampling strategy to match the versions of Python and NumPy. In addition, the special purpose jobs included are for linting, docs build, Gitpod build, Benchmarking, Pythran, MyPy, and more.
+
+## Miscellaneous
+
+In addition, I also worked on the upstream SciPy project on some of the minor issues. [PR #14993](https://github.com/scipy/scipy/pull/14493) updates all the Issue markdown templates to utilize the new YAML-based GitHub issue forms. The feedback received after the previous pull request was incorporated on [PR #14669](https://github.com/scipy/scipy/pull/14669).
+
+
+## Challenges faced
+
+While this whole project was challenging, some aspects of it were unforeseen and more challenging. When I decided to go on with this project, I had enough GitHub Actions knowledge to write the pipelines for small projects. I had not enough experience to work with complex GitHub Actions workflows on this big scale. With support from mentors, I was able to overcome this challenge with flying colors.
+
+The other challenge was understanding the Meson build system and how it is being used to build a complex project like SciPy with compiled dependencies. My experience with build systems was limited to `setuptools` and NPM for Python and JavaScript, respectively. After digging through a lot of extensive documentation, I figured out the nits and grits around the Meson build system.
+
+The most challenging task was to make the workflows work nicely with GitHub Actions. It was associated with many debugging sessions using [Tmate](https://github.com/marketplace/actions/debugging-with-tmate), which took me the most time. All thanks to my mentor John who helped me hack through the CI workflows and the build system in no time.
+
+## Future Plans
+
+There is a lot to do with the new CI matrix and the configurations, and it will take more time to reach the desired maturity. I was able to meet most of the goals, but some of them remain.
+
+- Implement the CI matrix on the GitHub Actions workflow.
+
+- Add associated developer documentation for CI matrix and service providers.
+
+- Improve and optimize the project as it is tested on multiple architectures.
+
+- Automate source and binary distribution through GitHub releases.
+
+## What did I learn from this project?
+
+The summer internship at Quansight Labs has been the busiest time of my life for all good reasons. I learned a great deal about GitHub Actions, build systems, and PyData projects through the same. I also learned about testing and benchmarking, along with Docker by a significant margin. I had the basic idea for some of them and have worked on them in hobby projects, but implementing them myself on a significant project and fixing bugs was a wholesome experience.
```

Comment:
the "by a significant margin" seems a little off to me. Maybe "along with a more in-depth understanding of Docker" or something like that?